src/jwt.ts

```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
+    throw new Error(
+      `Issuer (iss) claim mismatch in the ID token; expected "${options.iss}", found "${decoded.claims.iss}"`
+    );
   }
-  if (decoded.claims.aud !== options.aud) {
-    throw new Error('Invalid audience');
+
+  if (!decoded.user.sub) {
+    throw new Error(
+      'Subject (sub) claim must be a string present in the ID token'
+    );
   }
+
   if (decoded.header.alg !== 'RS256') {
-    throw new Error('Invalid algorithm');
+    throw new Error(
+      `Signature algorithm of "${decoded.header.alg}" is not supported. Expected the ID token to be signed with "RS256".`
+    );
+  }
+
+  if (
+    !decoded.claims.aud ||
+    !(
+      typeof decoded.claims.aud === 'string' ||
+      Array.isArray(decoded.claims.aud)
+    )
+  ) {
+    throw new Error(
+      'Audience (aud) claim must be a string or array of strings present in the ID token'
+    );
   }
+  if (Array.isArray(decoded.claims.aud)) {
+    if (!decoded.claims.aud.includes(options.aud)) {
+      throw new Error(
+        `Audience (aud) claim mismatch in the ID token; expected "${
+          options.aud
+        }" but was not one of "${decoded.claims.aud.join(', ')}"`
+      );
+    }
+    if (decoded.claims.aud.length > 1) {
+      if (!decoded.claims.azp) {
+        throw new Error(
+          'Authorized Party (azp) claim must be a string present in the ID token when Audience (aud) claim has multiple values'
+        );
+      }
+      if (decoded.claims.azp !== options.client_id) {
+        throw new Error(
+          `Authorized Party (azp) claim mismatch in the ID token; expected "${options.aud}", found "${decoded.claims.azp}"`
+        );
+      }
+    }
+  } else if (decoded.claims.aud !== options.aud) {
+    throw new Error(
+      `Audience (aud) claim mismatch in the ID token; expected "${options.aud}" but found "${decoded.claims.aud}"`
+    );
+  }
+
+  if (!decoded.claims.nonce) {
+    throw new Error(
+      'Nonce (nonce) claim must be a string present in the ID token'
+    );
+  }
+
   if (decoded.claims.nonce !== options.nonce) {
-    throw new Error('Invalid nonce');
+    throw new Error(
+      `Nonce (nonce) claim mismatch in the ID token; expected "${options.nonce}", found "${decoded.claims.nonce}"`
+    );
+  }
+
+  if (options.max_age && !decoded.claims.auth_time) {
+    throw new Error(
+      'Authentication Time (auth_time) claim must be a number present in the ID token when Max Age (max_age) is specified'
+    );
+  }
+
+  /* istanbul ignore next */
+  if (!decoded.claims.exp) {
+    throw new Error(
+      'Expiration Time (exp) claim must be a number present in the ID token'
+    );
+  }
+  if (!decoded.claims.iat) {
+    throw new Error(
+      'Issued At (iat) claim must be a number present in the ID token'
+    );
   }
 
   const now = new Date();
   const expDate = new Date(0);
   const iatDate = new Date(0);
   const nbfDate = new Date(0);
   const leeway = options.leeway || 60;
+  const authTimeDate = new Date(
```

Comment:
why is this one being instantiated like this and the rest are using the `setUTCSeconds`? Seems all should be using the latter

Reply:
fixed